File: studio/backend/core/inference/transcript_stream.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright 2026-present the Unsloth AI Inc. team. All rights reserved. See /studio/LICENSE.AGPL-3.0

"""Progress and durable results for the Studio transcription screen."""

import asyncio
import contextlib
import json
import time

from fastapi import HTTPException

from core.inference import transcript_gallery
from loggers import get_logger

logger = get_logger(__name__)


async def stream_transcript(transcribe, title: str):
    loop = asyncio.get_running_loop()
    updates = asyncio.Queue(maxsize=1)
    closed = False
    last_update = 0.0

    def publish(update):
        if closed:
            return
        if updates.full():
            updates.get_nowait()
        updates.put_nowait({"type": "progress", **update})

    def on_progress(update):
        nonlocal last_update
        if closed:
            return
        now = time.monotonic()
        if now - last_update >= 0.25:
            last_update = now
            loop.call_soon_threadsafe(publish, update)

    async def run():
        result = await transcribe(on_progress)
        record = None
        if result.get("text", "").strip():
            try:
                record = await asyncio.to_thread(transcript_gallery.save, result, title)
            except Exception as exc:
                logger.warning("Could not save transcript to the gallery (%s)", type(exc).__name__)
        return {"type": "complete", **result, "record": record}

    task = asyncio.create_task(run())
    pending = asyncio.create_task(updates.get())
    try:
        yield json.dumps({"type": "progress", "text": ""}) + "\n"
        while not task.done():
            done, _ = await asyncio.wait(
                {task, pending}, timeout=5, return_when=asyncio.FIRST_COMPLETED
            )
            if pending in done:
                yield json.dumps(pending.result()) + "\n"
                pending = asyncio.create_task(updates.get())
            elif not done:
                yield json.dumps({"type": "heartbeat"}) + "\n"
        yield json.dumps(task.result()) + "\n"
    except HTTPException as exc:
        yield json.dumps({"type": "error", "message": exc.detail}) + "\n"
    except Exception as exc:
        logger.warning("Transcription stream failed (%s)", type(exc).__name__)
        yield json.dumps({"type": "error", "message": "Transcription failed. Try again."}) + "\n"
    finally:
        closed = True
        pending.cancel()
        task.cancel()
        with contextlib.suppress(asyncio.CancelledError):
            await pending
        with contextlib.suppress(asyncio.CancelledError, Exception):
            await task

File: studio/frontend/src/features/audio/transcript-stream.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright 2026-present the Unsloth AI Inc. team. All rights reserved. See /studio/LICENSE.AGPL-3.0

export interface TranscriptRecord {
  id: string;
  title: string;
  text: string;
  model: string;
  language: string | null;
  duration: number | null;
  created_at: string;
  archived: boolean;
}

export interface TranscriptProgress {
  text: string;
  processed_seconds?: number;
  duration?: number;
}

export interface TranscriptResult {
  text: string;
  model: string;
  duration: number | null;
  record: TranscriptRecord | null;
}

export async function readTranscriptStream(
  body: ReadableStream<Uint8Array>,
  onProgress: (progress: TranscriptProgress) => void,
): Promise<TranscriptResult> {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  try {
    for (;;) {
      const { done, value } = await reader.read();
      buffer += decoder.decode(value, { stream: !done });
      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";
      if (done && buffer.trim()) lines.push(buffer);
      for (const line of lines) {
        if (!line.trim()) continue;
        const event = JSON.parse(line);
        if (event.type === "error") throw new Error(event.message);
        if (event.type === "progress") onProgress(event);
        if (event.type === "complete") return event as TranscriptResult;
      }
      if (done)
        throw new Error(
          "The transcription connection ended before the result arrived.",
        );
    }
  } finally {
    await reader.cancel().catch(() => undefined);
    reader.releaseLock();
  }
}
